File: scip-java/src/main/scala/com/sourcegraph/scip_java/ScipSymbol.scala
package com.sourcegraph.scip_java

import com.sourcegraph.scip.ScipSymbols
import com.sourcegraph.scip_aggregator.SymbolDescriptor

sealed abstract class ScipSymbol {}
final case class LocalScipSymbol(identifier: String) extends ScipSymbol
final case class GlobalScipSymbol(
    scheme: String,
    packageManager: String,
    packageName: String,
    packageVersion: String,
    descriptors: List[SymbolDescriptor]
) extends ScipSymbol

object ScipSymbol {

  def parseOrThrowExceptionIfInvalid(scipSymbol: String): ScipSymbol = {
    if (scipSymbol.startsWith("local ")) {
      LocalScipSymbol(scipSymbol.stripPrefix("local "))
    } else {
      scipSymbol.split(" ", 5) match {
        case Array(
              scheme,
              packageManager,
              packageName,
              packageVersion,
              descriptor
            ) =>
          GlobalScipSymbol(
            scheme,
            packageManager,
            packageName,
            packageVersion,
            parseDescriptors(descriptor)
          )
        case _ =>
          throw new IllegalArgumentException(
            s"Invalid scip symbol: $scipSymbol"
          )
      }
    }
  }

  private def parseDescriptors(scipSymbol: String): List[SymbolDescriptor] = {
    val descriptor = SymbolDescriptor.parseFromSymbol(scipSymbol)
    if (descriptor.owner == ScipSymbols.ROOT_PACKAGE)
      Nil
    else
      descriptor :: parseDescriptors(descriptor.owner)
  }

}

File: scip-java/src/main/kotlin/com/sourcegraph/scip_java/ScipSymbol.kt
package com.sourcegraph.scip_java

import com.sourcegraph.scip.ScipSymbols
import com.sourcegraph.scip_aggregator.SymbolDescriptor

/**
 * Validates that a string is a syntactically valid SCIP symbol.
 *
 * The snapshot printer uses this to fail fast when an index contains a symbol
 * that doesn't parse. The parsed structure isn't needed by any caller, so only
 * the throw-on-invalid behaviour is kept.
 */
object ScipSymbol {
    fun parseOrThrowExceptionIfInvalid(scipSymbol: String) {
        if (scipSymbol.startsWith("local ")) return
        val parts = scipSymbol.split(" ", limit = 5)
        require(parts.size == 5) { "Invalid scip symbol: $scipSymbol" }
        var current = parts[4]
        while (true) {
            val descriptor = SymbolDescriptor.parseFromSymbol(current)
            if (descriptor.owner == ScipSymbols.ROOT_PACKAGE) return
            current = descriptor.owner
        }
    }
}
